Explore a fix for broken casava format handling in a number of scripts

ctb opened the pull request:
See #817, among others.

ctb commented:
Confronting #817 head-on, we have a few different problems:

- split-paired-reads.py would drop reads that it couldn't interpret (oops)
- screed by default splits up the record name into 'name' and 'description', which breaks Casava 1.8 names
- this splitting leads to silly results -- e.g. see https://github.com/ged-lab/khmer/blob/fix/casava_formats/tests/test_scripts.py#L2316

It looks to me like the right thing to do is make sure that parse_description is always set to false
in screed iterators. This will necessitate changing about a dozen things in scripts/. @mr-c thoughts on this wrt semantic versioning?

cc @standage

ctb commented:
Addresses #763

standage commented:
This fixes the splitting/interleaving issues with Casava 1.8 AFAICS.

mr-c commented:
The changes might not be fun but if the tests pass without modifications then we are violating the semantic versioning.

ctb commented:
Sorry, clarify: yes, fix, and to heck with semver?


> On Feb 21, 2015, at 8:26 PM, Michael R. Crusoe <notifications@github.com> wrote:
> 
> The changes might not be fun but if the tests pass without modifications then we are violating the semantic versioning.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mr-c commented:
mea culpa; ----> Then we are _not_ violating semantic versioning if the
tests continue to pass unmodified.

On Sat Feb 21 2015 at 8:34:43 PM C. Titus Brown <notifications@github.com>
wrote:

> Sorry, clarify: yes, fix, and to heck with semver?
>
>
> > On Feb 21, 2015, at 8:26 PM, Michael R. Crusoe <notifications@github.com>
> wrote:
> >
> > The changes might not be fun but if the tests pass without modifications
> then we are violating the semantic versioning.
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ged-lab/khmer/pull/818#issuecomment-75412970>.
>

mr-c commented:
To summarize my understanding: khmer has always mangled sequence record names: either by dropping descriptions or appending different pairing notation on the end.

This PR aims to bring khmer in line with expected practice: both old and new style pair naming are accepted everywhere; descriptions will no longer be dropped.

With respect to semantic versioning we have three options:

A) We declare the old behavior a horrible bug and widely publicize the correction and thus we don't increment the major version number.

or 

B) We accept that the old behavior was present for such a long time and likely required workarounds that will break upon its fixing. Therefore we increment the major version to indicate this, thus following our commitment to semantic (and not 'marketing') versioning.

or 

C) we hid the corrected behavior behind a switch and wait for the 2.0 release to make it the default behavior

As Titus privately pointed out to me, seeing what changes will be required to update the protocols (if anything) will inform the wisdom of option B or C.

ctb commented:
While additional tests are needed, and it's not yet ready for a full CR, it's mostly done, I think? Thoughts welcome.  I'll finish sanding down the corners tomorrow or later today, as time permits.

ctb commented:
- [ ] add tests for filter-abund
- [ ] add tests for filter-abund-single
- [ ] resolve issues with count-median now allowing spaces in output
- [ ] add tests for do-partition
- [ ] add tests for annotate-partition and extract-partition
- [ ] add tests for filter-stoptags
- [ ] add tests for normalize-by-median
- [ ] add tests for extract-long-sequences
- [ ] add tests for fastq-to-fasta
- [ ] add tests for sample-reads-randomly

OK, so it's a slightly bigger issue than I thought :).

ctb commented:
- [ ] file issue to update file formats to proper CSV, with headers, for khmer 2.0.

ctb commented:
- [ ] file an issue regarding --csv for anything that outputs tabular format; also see #789

ctb commented:
For posterity:

* fixed bug in utils.check_is_pair, introduced last week; https://github.com/ged-lab/khmer/blob/fix/casava_formats/khmer/utils.py#L36
* fixed old bug in split-paired-reads.py where reads with unknown pairing format would just get dropped: https://github.com/ged-lab/khmer/blob/fix/casava_formats/scripts/split-paired-reads.py#L143
* fixed bug in tests for extract-long-reads where output file was improperly named and not actually tested anyway: https://github.com/ged-lab/khmer/blob/fix/casava_formats/tests/test_scripts.py#L1995

So that was awesome... :)

ctb commented:
For me to do:

- [ ] do diff cover and diff pylint analysis.
- [ ] look at split-paired-reads.py and interleave-reads.py critically, to see if more tests are needed.

@mr-c and others, a code review would be welcome of points not related to the above checklist; basically, any obvious flaws in the approach, or concerns that I've missed? TIA!

mr-c commented:
Jenkins, test this please

ctb commented:
- [ ] Is it mergable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage.
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

ctb commented:
Ready for review, w00t.

ctb commented:
Additional issues/bugs found & fixed:

* removed no longer needed test-data files;
* interleave-reads.py used to happily interleave the same file twice; now it errors out.

mr-c commented on scripts/count-median.py:
```diff
@@ -34,8 +36,18 @@ def get_parser():
-    length.
+    The output file contains sequence id, median, average, stddev, and
+    seq length; fields are separated by spaces.  For khmer 1.x,
+    sequence names are split at the first space, which means that some
```
'For khmer 1.x count-median.py will split sequence names at the first space which means that some ..'

mr-c commented on scripts/count-median.py:
```diff
@@ -34,8 +36,18 @@ def get_parser():
+    uninformative names.  Use :option:`--csv` to fix this behavior.
+
+    :option:`--csv` will output a comma-separated value file with column
+    headers, and the entire sequence name will be kept.
```
45-46 is a bit redundant; especially when viewed in context: http://athyra.ged.msu.edu/~mcrusoe/kdocs/user/scripts.html#count-median-py

ctb replied:
I think the redundancy is not harmful, in this case, and may be helpful to people who only skim docs.

mr-c commented on tests/test_scripts.py:
```diff
@@ -258,10 +258,33 @@ def test_filter_abund_3_fq_retained():
+    infile = utils.get_temp_filename('test.fq')
+    in_dir = os.path.dirname(infile)
+
+    shutil.copyfile(utils.get_test_data('test-abund-read-2.paired2.fq'),
```
It would be nifty if Casava 1.8 formatted test sequence files had 'c18' in their name.

mr-c commented:
LGTM!

ctb commented:
w00t!
